rmw_dds_common/test/test_graph_cache.cpp

```diff
@@ -0,0 +1,1066 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <cstring>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "rmw_dds_common/graph_cache.hpp"
+
+using rmw_dds_common::GraphCache;
+
+// namespace, name
+using NamesAndNamespaces = std::vector<std::pair<std::string, std::string>>;
+
+void
+check_names_and_namespace(
+  const rcutils_string_array_t & names,
+  const rcutils_string_array_t & namespaces,
+  const NamesAndNamespaces & expected)
+{
+  ASSERT_EQ(names.size, namespaces.size);
+  ASSERT_EQ(expected.size(), names.size);
+  for (size_t i = 0; i < expected.size(); i++) {
+    ASSERT_THAT(
+      expected,
+      testing::Contains(testing::Pair(namespaces.data[i], names.data[i])));
+  }
+}
+
+// topic_name, vector of type names
+using NamesAndTypes =
+  std::vector<std::pair<std::string, std::vector<std::string>>>;
+
+void check_names_and_types(
+  const rmw_names_and_types_t & names_and_types,
+  const NamesAndTypes & expected)
+{
+  ASSERT_EQ(names_and_types.names.size, expected.size());
+
+  for (size_t i = 0; i < expected.size(); i++) {
+    const auto & item = expected[i];
+    EXPECT_EQ(std::get<0>(item), names_and_types.names.data[i]);
+    const auto & expected_types = std::get<1>(item);
+    const auto & types = names_and_types.types[i];
+    ASSERT_EQ(expected_types.size(), types.size);
+    for (size_t j = 0; j < expected_types.size(); j++) {
+      EXPECT_EQ(expected_types[j], types.data[j]);
+    }
+  }
+}
+
+std::string
+identity_demangle(const std::string & name)
+{
+  return name;
+}
```

Comment:
@ivanpauno hmm, I think I've seen this before :)

Maybe some common (de)mangling functionality should live here in `rmw_dds_common`.

Reply:
Sounds reasonable.

Reply:
I was thinking that all DDS mangle/demangle rules should be moved to a common packages.
Currently, each dds based rmw implementation (connext, fastrtps, etc) have their own functions to mangle/demangle ros names.
In that package, we should have an `identity_demangle` too.

If it sounds resonable, that can be done in a follow-up PR.

Reply:
Ok, fair enough.